src/crypto/algorithms/megolm.ts

```diff
@@ -1419,154 +1450,326 @@ export class MegolmDecryption extends DecryptionAlgorithm {
+
+        if (!claimedEd25519Key) {
+            logger.error(`forwarded_room_key_event is missing sender_claimed_ed25519_key field`);
+            return;
+        }
 
-            // replace the sender key with the sender key of the session
-            // creator for storage
-            senderKey = content.sender_key;
+        const keysClaimed = {
+            ed25519: claimedEd25519Key,
+        };
+
+        // FIXME: We're reusing the same field to track both:
+        //
+        // 1. The Olm identity we've received this room key from.
+        // 2. The Olm identity deduced (in the trusted case) or claiming (in the
+        // untrusted case) to be the original creator of this room key.
+        //
+        // We now overwrite the value tracking usage 1 with the value tracking usage 2.
+        roomKey.senderKey = claimedCurve25519Key;
+        // Replace our keysClaimed as well.
+        roomKey.keysClaimed = keysClaimed;
+        roomKey.exportFormat = true;
+        roomKey.forwardingKeyChain = forwardingKeyChain;
+        // forwarded keys are always untrusted
+        roomKey.extraSessionData.untrusted = true;
+
+        return roomKey;
+    }
+
+    /**
+     * Determine if we should accept the forwarded room key that was found in the given
+     * event.
+     *
+     * @param event - An `m.forwarded_room_key` event.
+     * @param roomKey - The room key that was found in the event.
+     *
+     * @returns promise that will resolve to a boolean telling us if it's ok to
+     * accept the given forwarded room key.
+     *
+     * @internal
+     *
+     */
+    private async shouldAcceptForwardedKey(event: MatrixEvent, roomKey: RoomKey): Promise<boolean> {
+        const senderKey = event.getSenderKey()!;
+
+        const sendingDevice =
+            this.crypto.deviceList.getDeviceByIdentityKey(olmlib.OLM_ALGORITHM, senderKey) ?? undefined;
+        const deviceTrust = this.crypto.checkDeviceInfoTrust(event.getSender()!, sendingDevice);
+
+        // Using the plaintext sender here is fine since we checked that the
+        // sender matches to the user id in the device keys when this event was
+        // originally decrypted. This can obviously only happen if the device
+        // keys have been downloaded, but if they haven't the
+        // `deviceTrust.isVerified()` flag would be false as well.
+        //
+        // It would still be far nicer if the `sendingDevice` had a user ID
+        // attached to it that went through signature checks.
+        const fromUs = event.getSender() === this.baseApis.getUserId();
+        const keyFromOurVerifiedDevice = deviceTrust.isVerified() && fromUs;
+        const weRequested = await this.wasRoomKeyRequested(event, roomKey);
+        const fromInviter = this.wasRoomKeyForwardedByInviter(event, roomKey);
+        const sharedAsHistory = this.wasRoomKeyForwardedAsHistory(roomKey);
+
+        return (weRequested && keyFromOurVerifiedDevice) || (fromInviter && sharedAsHistory);
```

Comment:
Another thing that might be clearer here is to break this out, allowing the addition of some comments:

```suggestion
        // we trust the forwarded key if we asked for it and it's from one of our own devices...
        if (weRequested && keyFromOurVerifiedDevice) {
            return true;
        }
        
        // ... or if it's from the user that invited us to the room, and it is being forwarded as part 
        // of the "shared history" of the room ...
        if (fromInviter && sharedAsHistory) {
            return true;
        }
        
        // ... otherwise, we ignore it.
        return false;
```

YMMV though; keep what you have if you prefer.

Reply:
I prefer the one-liner, at least unless more cases were to be added.